server/modules/routing/readwritesplit/rwsplit_route_stmt.cc

```diff
@@ -1107,6 +1108,60 @@ static inline bool is_large_query(GWBUF* buf)
     return buflen == MYSQL_HEADER_LEN + GW_MYSQL_MAX_PACKET_LEN;
 }
 
+/*
+ * Add a wait gitd query in front of user's query to achive causal read;
+ *
+ * @param inst   RWSplit
+ * @param rses   RWSplitSession
+ * @param server SERVER
+ * @param origin origin send buffer
+ * @return       A new buffer contains wait statement and origin query
+ */
+GWBUF *add_prefix_wait_gtid(RWSplit *inst, RWSplitSession *rses, SERVER *server, GWBUF *origin)
+{
+
+    /**
+     * Pack wait function and client query into a multistatments will save a round trip latency,
+     * and prevent the client query being executed on timeout.
+     * For example:
+     * SET @maxscale_secret_variable=(SELECT CASE WHEN MASTER_GTID_WAIT('232-1-1', 10) = 0
+     * THEN 1 ELSE (SELECT 1 FROM INFORMATION_SCHEMA.ENGINES) END); SELECT * FROM `city`;
+     * when MASTER_GTID_WAIT('232-1-1', 0.05) == 1 (timeout), it will return
+     * an error, and SELECT * FROM `city` will not be executed, then we can retry
+     * on master;
+     **/
+
+    GWBUF *rval;
+    const char* wait_func = (server->server_type == SERVER_TYPE_MARIADB) ?
+            MARIADB_WAIT_GTID_FUNC : MYSQL_WAIT_GTID_FUNC;
+    const char *gtid_wait_timeout = inst->config().causal_read_timeout.c_str();
+    const char *gtid_pos = rses->gtid_pos.c_str();
+
+    /* Create a new buffer to store prefix sql */
+    size_t prefix_len = strlen(gtid_wait_stmt) + strlen(gtid_pos) +
+        strlen(gtid_wait_timeout) + strlen(wait_func);
+    char prefix_sql[prefix_len];
+    snprintf(prefix_sql, prefix_len, gtid_wait_stmt, wait_func, gtid_pos, gtid_wait_timeout);
+    GWBUF *prefix_buff = modutil_create_query(prefix_sql);
+
+    /**
+     * Trim origin to sql, Append origin buffer to the prefix buffer
+     **/
+    uint8_t header[MYSQL_HEADER_LEN];
+    gwbuf_copy_data(origin, 0, MYSQL_HEADER_LEN, header);
+    /* Command length = 1 */
+    size_t origin_sql_len = gw_mysql_get_byte3(header) - 1;
+    /* Trim mysql header and command */
+    GWBUF_LTRIM(origin, MYSQL_HEADER_LEN + 1);
```

Comment:
This can be replaced with `gwbuf_consume` which also correctly handles headers split between two buffers.